text/0156-columnar-encoding.md

```diff
@@ -0,0 +1,1417 @@
+This behavior is described in more details in the section [Best effort delivery guarantee](#best-effort-delivery-guarantee).
+
+The protobuf definition of this service is:
+
+```protobuf
+service ArrowStreamService {
+  // The ArrowStream endpoint is a bi-directional stream used to send batch of `BatchArrowRecords` from the exporter
+  // to the collector. The collector returns `BatchStatus` messages to acknowledge the `BatchArrowRecords` messages received.
+  rpc ArrowStream(stream BatchArrowRecords) returns (stream BatchStatus) {}
+}
+```
+
+> **Unary RPC vs Stream RPC**: We use a stream-oriented protocol **to get rid of the overhead of specifying the schema
+> and dictionaries for each batch.** A state will be maintained receiver side to keep track of the schemas and
+> dictionaries. The [Arrow IPC format](#arrow-ipc-format) has been designed to follow this pattern and also allows the
+> dictionaries to be sent incrementally. Similarly, ZSTD dictionaries can also be transferred to the RPC stream to
+> optimize the transfer of small batches (for more details see the description of the compression field in the
+> following paragraphs). To mitigate the usual pitfalls of a stream-oriented protocol please see this [paragraph](#traffic-balancing-optimization) in the
+> implementation recommendations section.
+
+A `BatchArrowRecords` message is composed of 3 attributes. The protobuf definition is:
+
+```protobuf
+message BatchArrowRecords {
+  // [mandatory] Batch ID. Must be unique in the context of the stream.
+  string batch_id = 1;
+
+  // [mandatory] A collection of payloads containing the data of the batch.
+  repeated OtlpArrowPayload otlp_arrow_payloads = 2;
+}
+```
+
+The `batch_id` attribute is a unique identifier for the batch inside the scope of the current stream. It is used to
+uniquely
+identify the batch in the egress `BatchStatus` stream. See the [Batch Id generation](#batch-id-generation) section for
+more information on the implementation of this identifier.
+
+The `otlp_arrow_payloads` attribute is a list of `OtlpArrowPayload` messages. Each `OtlpArrowPayload` message represents
+a table of data encoded in a columnar format (e.g. metrics, logs, traces, and future OTLP entities). Several correlated
+IPC Arrow messages of different nature and with different schemas can be sent in the same OTLP batch identified by
+`batch_id` and thus be processed as one unit without complex logic in the collector or any other processing systems.
+More details on the `OtlpArrowPayload` columns in the section [Mapping OTEL entities to Arrow records](#mapping-otel-entities-to-arrow-records).
+
+More specifically, an `OtlpArrowPayload` protobuf message is defined as:
+
+```protobuf
+message OtlpArrowPayload {
+  // [mandatory] A unique id assigned to a sub-stream of the batch sharing the same schema, and dictionaries.
+  string sub_stream_id = 1;
```

Comment:
It looks like we are building our own sub-streams inside a streams, which is additional complication. Why not open multiple gRPC streams instead to make implementation simpler?

Reply:
Do multiple sub-stream have the same dictionary or different dictionaries? The word "sharing" is confusing.